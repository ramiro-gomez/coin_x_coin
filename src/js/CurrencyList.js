import Currency from './Currency';
import Conversor from './Conversor';

export default class CurrencyList {
	#currencies;

	constructor({ currencyNames, exchangeRates, base }) {
		this.#currencies = currencyNames.reduce((acc, currency) => {
			//  Avoid metals
			if (currency.name === 'Silver (Troy Ounce)'
				|| currency.name === 'Gold (Troy Ounce)'
				|| currency.name === 'Palladium'
				|| currency.name === 'Platinum') {
				return { ...acc };
			}
			return {
				...acc,
				[currency.id]: new Currency({
					id: currency.id,
					name: currency.name,
					price: {
						value: Conversor.toBase(exchangeRates[currency.id]),
						base,
					},
				}),
			};
		}, {});
	}

	get currencies() {
		return this.#currencies;
	}
}
