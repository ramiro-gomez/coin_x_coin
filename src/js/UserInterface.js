export default class UserInterface {
	static showCards(currencyList) {
		const documentFragment = new DocumentFragment();
		const middleElement = document.querySelector('.middle-container');
		Object.values(currencyList.currencies).forEach((currency) => {
			const card = document.createElement('div');
			card.classList.add('card');
			card.innerHTML = `
				<button type="button">
					<i class="fas fa-thumbtack"></i>
				</button>
				<div class="card-name">
					<h2>${currency.id}</h2>
					<p>${currency.name}</p>
				</div>
				<div class="card-price">
					<p>1<span>${currency.id}</span> = </p>
					<h3>${currency.price.value}<span>${currency.price.base}</span></h3>
				</div>
			`;
			documentFragment.appendChild(card);
		});
		middleElement.appendChild(documentFragment);
	}

	static updateCardPrices(currencyList) {
		const cardElements = document.querySelectorAll('.middle-container div.card');
		cardElements.forEach((card) => {
			const id = card.querySelector('.card-name h2').textContent;
			if (currencyList.currencies[id]) {
				const priceValue = card.querySelector('.card-price h3');
				const priceBasis = card.querySelector('.card-price h3 span');
				priceValue.childNodes[0].nodeValue = currencyList.currencies[id].price.value;
				priceBasis.textContent = currencyList.currencies[id].price.base;
			}
		});
	}

	static toggleCardPushpin(card) {
		const pushpin = card.querySelector('button i');
		pushpin.classList.toggle('pushpin');
		card.classList.toggle('order-first');
	}

	static addsavedPushpins = ({ toCardElements, savedPushpins }) => {
		toCardElements.forEach((card) => {
			const id = card.querySelector('div.card-name h2').textContent;
			if (savedPushpins.includes(id)) {
				card.querySelector('i.fa-thumbtack').classList.add('pushpin');
				card.classList.add('order-first');
			}
		});
	}

	static showOptions({
		currencyList, savedBase, savedExchange0, savedExchange1,
	}) {
		const baseSelector = document.querySelector('div.base-bar select');
		const exchangeSelector = document.querySelectorAll('div.exchange-box select');
		const baseDocumentFrag = new DocumentFragment();
		Object.values(currencyList.currencies).forEach((currency) => {
			const option = document.createElement('option');
			option.value = currency.id;
			option.textContent = currency.id;
			baseDocumentFrag.appendChild(option);
		});

		const exchangeDocumentFrag0 = baseDocumentFrag.cloneNode(true);
		const exchangeDocumentFrag1 = baseDocumentFrag.cloneNode(true);

		if (savedBase) baseDocumentFrag.querySelector(`option[value=${savedBase}]`).selected = true;
		if (savedExchange0) exchangeDocumentFrag0.querySelector(`option[value=${savedExchange0}]`).selected = true;
		if (savedExchange1) exchangeDocumentFrag1.querySelector(`option[value=${savedExchange1}]`).selected = true;

		baseSelector.appendChild(baseDocumentFrag);
		exchangeSelector[0].appendChild(exchangeDocumentFrag0);
		exchangeSelector[1].appendChild(exchangeDocumentFrag1);
	}

	static showValueInExchangeInput({ showInInputNumber, valueToShow }) {
		const exchangeInput = document.querySelectorAll('div.exchange-box input');
		exchangeInput[showInInputNumber].value = valueToShow;
	}

	static showModal() {
		const modalContainer = document.querySelector('.modal-container');
		modalContainer.classList.add('open');
	}

	static hideModal() {
		const modalContainer = document.querySelector('.modal-container');
		modalContainer.classList.remove('open');
	}
}
