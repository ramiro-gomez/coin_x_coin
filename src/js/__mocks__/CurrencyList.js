import Currency from '../Currency';

export default class CurrencyList {
	#currencies = {
		AED: new Currency({
			id: 'AED',
			name: 'United Arab Emirates Dirham',
			price: {
				value: 10,
				base: 'ARS',
			},
		}),
		AFN: new Currency({
			id: 'AFN',
			name: 'Afghan Afghani',
			price: {
				value: 25,
				base: 'ARS',
			},
		}),
		ALL: new Currency({
			id: 'ALL',
			name: 'Albanian Lek',
			price: {
				value: 5,
				base: 'ARS',
			},
		}),
		AMD: new Currency({
			id: 'AMD',
			name: 'Armenian Dram',
			price: {
				value: 30,
				base: 'ARS',
			},
		}),
	};

	get currencies() {
		return this.#currencies;
	}
}
