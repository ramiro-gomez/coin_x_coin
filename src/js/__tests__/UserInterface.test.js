import fs from 'fs';
import path from 'path';
import { JSDOM } from '../../../node_modules/jsdom';
import UserInterface from '../UserInterface';
import CurrencyList from '../CurrencyList';

jest.mock('../CurrencyList.js');
const mockCurrencyList = new CurrencyList();

const HTML = fs.readFileSync(path.resolve(__dirname, '../../index.html'), 'utf8');
const DOM = new JSDOM(HTML);
document.body.innerHTML = DOM.window.document.body.innerHTML;
const mockModalContainer = DOM.window.document.querySelector('.modal-container').outerHTML;
const mockCards = `
	<div class="card">
		<button type="button">
			<i class="fas fa-thumbtack"></i>
		</button>
		<div class="card-name">
			<h2>AED</h2>
			<p>United Arab Emirates Dirham</p>
		</div>
		<div class="card-price">
			<p>1<span>AED</span> = </p>
			<h3>25.9<span>ARS</span></h3>
		</div>
	</div>
	<div class="card">
		<button type="button">
			<i class="fas fa-thumbtack"></i>
		</button>
		<div class="card-name">
			<h2>AFN</h2>
			<p>Afghan Afghani</p>
		</div>
		<div class="card-price">
			<p>1<span>AFN</span> = </p>
			<h3>1.21<span>ARS</span></h3>
		</div>
	</div>
	<div class="card">
		<button type="button">
			<i class="fas fa-thumbtack"></i>
		</button>
		<div class="card-name">
			<h2>ALL</h2>
			<p>Albanian Lek</p>
		</div>
		<div class="card-price">
			<p>1<span>ALL</span> = </p>
			<h3>0.94<span>ARS</span></h3>
		</div>
	</div>
	<div class="card">
		<button type="button">
			<i class="fas fa-thumbtack"></i>
		</button>
		<div class="card-name">
			<h2>AWG</h2>
			<p>Aruban Florin</p>
		</div>
		<div class="card-price">
			<p>1<span>AWG</span> = </p>
			<h3>3.5<span>ARS</span></h3>
		</div>
	</div>
`;

describe('#showCards', () => {
	it('should render the cards in the middle-container', () => {
		const middleElement = document.querySelector('.middle-container');
		UserInterface.showCards(mockCurrencyList);
		expect(middleElement).toMatchSnapshot();
	});
});

describe('#updateCardPrices', () => {
	beforeEach(() => {
		document.querySelector('.middle-container').innerHTML = mockCards;
	});
	it('should update the prices of the cards with the given currencies', () => {
		const cardElements = document.querySelectorAll('.middle-container div.card');
		UserInterface.updateCardPrices(mockCurrencyList);
		let assertions = 0;
		cardElements.forEach((card) => {
			const id = card.querySelector('.card-name h2').textContent;
			const priceValue = card.querySelector('.card-price h3').childNodes[0].nodeValue.replace(/\s+/g, '');
			const priceBasis = card.querySelector('.card-price h3 span').textContent;
			if (mockCurrencyList.currencies[id]) {
				expect(priceValue).toBe(mockCurrencyList.currencies[id].price.value.toString());
				expect(priceBasis).toBe(mockCurrencyList.currencies[id].price.base);
				assertions += 2;
			}
		});
		expect.assertions(assertions);
	});
});

describe('#toggleCardPushpin', () => {
	beforeEach(() => {
		document.querySelector('.middle-container').innerHTML = mockCards;
	});
	afterAll(() => {
		document.querySelector('.middle-container').innerHTML = '';
	});
	it('should toggle the classes pushpin and order-first of the given card', () => {
		const card = document.querySelector('.middle-container .card');
		const pushpin = card.querySelector('button i');
		UserInterface.toggleCardPushpin(card);
		expect(pushpin.className).toMatch('pushpin');
		expect(card.className).toMatch('order-first');
		UserInterface.toggleCardPushpin(card);
		expect(pushpin.className).not.toMatch('pushpin');
		expect(card.className).not.toMatch('order-first');
	});
});

describe('#addDefaultPushpins', () => {
	beforeEach(() => {
		document.querySelector('.middle-container').innerHTML = mockCards;
	});
	afterAll(() => {
		document.querySelector('.middle-container').innerHTML = '';
	});
	it('should add the pushpin and order-first class to the cards of the indicated currencies', () => {
		const mockDefaultPushpins = ['AFN', 'ALL'];
		const cardElements = document.querySelectorAll('.middle-container div.card');
		UserInterface.addDefaultPushpins({
			toCardElements: cardElements,
			defaultPushpins: mockDefaultPushpins,
		});
		cardElements.forEach((card) => {
			const id = card.querySelector('div.card-name h2').textContent;
			const pushpinIcon = card.querySelector('button i');
			if (mockDefaultPushpins.includes(id)) {
				expect(pushpinIcon.classList.contains('pushpin')).toBeTruthy();
				expect(card.classList.contains('order-first')).toBeTruthy();
			} else {
				expect(pushpinIcon.classList.contains('pushpin')).toBeFalsy();
				expect(card.classList.contains('order-first')).toBeFalsy();
			}
		});
		expect.assertions(cardElements.length * 2);
	});
});

describe('#showOptions', () => {
	afterEach(() => {
		document.querySelector('div.base-bar select').innerHTML = '';
	});
	it('should render the options of the base selector and exchange input selector 0 and 1', () => {
		const baseSelector = document.querySelector('div.base-bar select');
		const exchangeSelector = document.querySelectorAll('div.exchange-box select');
		UserInterface.showOptions({
			currencyList: mockCurrencyList,
		});
		expect(baseSelector).toMatchSnapshot();
		expect(exchangeSelector[0]).toMatchSnapshot();
		expect(exchangeSelector[1]).toMatchSnapshot();
	});
	it('should select the default option of the base selector and exchange input selector 0 and 1', () => {
		const baseSelector = document.querySelector('div.base-bar select');
		const exchangeSelector = document.querySelectorAll('div.exchange-box select');
		UserInterface.showOptions({
			currencyList: mockCurrencyList,
			defaultBase: 'AFN',
			defaultExchange0: 'ALL',
			defaultExchange1: 'AMD',
		});
		expect(baseSelector.value).toBe('AFN');
		expect(exchangeSelector[0].value).toBe('ALL');
		expect(exchangeSelector[1].value).toBe('AMD');
	});
});

describe('#showValueInExchangeInput', () => {
	afterAll(() => {
		document.querySelectorAll('div.exchange-box input').forEach((input) => {
			input.value = '';
		});
	});
	it('should render the value in the exchange input 0', () => {
		const exchangeInput = document.querySelectorAll('div.exchange-box input');
		UserInterface.showValueInExchangeInput({
			showInInputNumber: 0,
			valueToShow: 15,
		});
		expect(exchangeInput[0].value).toBe('15');
	});
	it('should render the value in the exchange input 1', () => {
		const exchangeInput = document.querySelectorAll('div.exchange-box input');
		UserInterface.showValueInExchangeInput({
			showInInputNumber: 1,
			valueToShow: 30,
		});
		expect(exchangeInput[1].value).toBe('30');
	});
});

describe('#showModal', () => {
	afterEach(() => {
		document.querySelector('.modal-container').outerHTML = mockModalContainer;
	});
	it('should add the class open to modal-container', () => {
		const modalContainer = document.querySelector('.modal-container');
		UserInterface.showModal();
		expect(modalContainer.classList.contains('open')).toBeTruthy();
	});
});

describe('#hideModal', () => {
	afterEach(() => {
		document.querySelector('.modal-container').outerHTML = mockModalContainer;
	});
	it('should remove the class open to modal-container', () => {
		const modalContainer = document.querySelector('.modal-container');
		modalContainer.classList.add('open');
		UserInterface.hideModal();
		expect(modalContainer.classList.contains('open')).toBeFalsy();
	});
});
