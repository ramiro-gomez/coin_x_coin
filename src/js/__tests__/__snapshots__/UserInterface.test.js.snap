// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#showCards should render the cards in the middle-container 1`] = `
<section
  class="middle-container"
>
  
        
  <div
    class="card"
  >
    
				
    <button
      type="button"
    >
      
					
      <i
        class="fas fa-thumbtack"
      />
      
				
    </button>
    
				
    <div
      class="card-name"
    >
      
					
      <h2>
        AED
      </h2>
      
					
      <p>
        United Arab Emirates Dirham
      </p>
      
				
    </div>
    
				
    <div
      class="card-value"
    >
      
					
      <p>
        1
        <span>
          AED
        </span>
         = 
      </p>
      
					
      <h3>
        10
        <span>
          ARS
        </span>
      </h3>
      
				
    </div>
    
			
  </div>
  <div
    class="card"
  >
    
				
    <button
      type="button"
    >
      
					
      <i
        class="fas fa-thumbtack"
      />
      
				
    </button>
    
				
    <div
      class="card-name"
    >
      
					
      <h2>
        AFN
      </h2>
      
					
      <p>
        Afghan Afghani
      </p>
      
				
    </div>
    
				
    <div
      class="card-value"
    >
      
					
      <p>
        1
        <span>
          AFN
        </span>
         = 
      </p>
      
					
      <h3>
        25
        <span>
          ARS
        </span>
      </h3>
      
				
    </div>
    
			
  </div>
  <div
    class="card"
  >
    
				
    <button
      type="button"
    >
      
					
      <i
        class="fas fa-thumbtack"
      />
      
				
    </button>
    
				
    <div
      class="card-name"
    >
      
					
      <h2>
        ALL
      </h2>
      
					
      <p>
        Albanian Lek
      </p>
      
				
    </div>
    
				
    <div
      class="card-value"
    >
      
					
      <p>
        1
        <span>
          ALL
        </span>
         = 
      </p>
      
					
      <h3>
        5
        <span>
          ARS
        </span>
      </h3>
      
				
    </div>
    
			
  </div>
  <div
    class="card"
  >
    
				
    <button
      type="button"
    >
      
					
      <i
        class="fas fa-thumbtack"
      />
      
				
    </button>
    
				
    <div
      class="card-name"
    >
      
					
      <h2>
        AMD
      </h2>
      
					
      <p>
        Armenian Dram
      </p>
      
				
    </div>
    
				
    <div
      class="card-value"
    >
      
					
      <p>
        1
        <span>
          AMD
        </span>
         = 
      </p>
      
					
      <h3>
        30
        <span>
          ARS
        </span>
      </h3>
      
				
    </div>
    
			
  </div>
</section>
`;

exports[`#showOptions should render the options of the base selector and exchange input selector 0 and 1 1`] = `
<select>
  <option
    value="AED"
  >
    AED
  </option>
  <option
    value="AFN"
  >
    AFN
  </option>
  <option
    value="ALL"
  >
    ALL
  </option>
  <option
    value="AMD"
  >
    AMD
  </option>
</select>
`;

exports[`#showOptions should render the options of the base selector and exchange input selector 0 and 1 2`] = `
<select>
  <option
    value="AED"
  >
    AED
  </option>
  <option
    value="AFN"
  >
    AFN
  </option>
  <option
    value="ALL"
  >
    ALL
  </option>
  <option
    value="AMD"
  >
    AMD
  </option>
</select>
`;

exports[`#showOptions should render the options of the base selector and exchange input selector 0 and 1 3`] = `
<select>
  <option
    value="AED"
  >
    AED
  </option>
  <option
    value="AFN"
  >
    AFN
  </option>
  <option
    value="ALL"
  >
    ALL
  </option>
  <option
    value="AMD"
  >
    AMD
  </option>
</select>
`;
